a10_octavia/common/utils.py

```diff
@@ -40,32 +40,43 @@ def validate_params(rack_info):
                                                   'username', 'password', 'device_name')):
             if validate_ipv4(rack_info['ip_address']):
                 return True
-            LOG.error('Invalid IP address given ' + rack_info['ip_address'])
-    LOG.error('Configuration of `devices` under [RACK_VTHUNDER] is invalid. '
-              'Please check your configuration. The params `project_id`, '
-              '`ip_address`, `username`, `password` and `device_name` cannot be None ')
+            raise ConfigFileValueError('Invalid IPAddress value given ' + rack_info['ip_address'])
+    raise ConfigFileValueError('Please check your configuration. The params `project_id`, '
+                               '`ip_address`, `username`, `password` and `device_name` '
+                               'under [rack_vthunder] section cannot be None ')
     return False
 
 
+def check_duplicate_entries(rack_dict):
+    rack_count_dict = {}
+    for rack_key, rack_value in rack_dict.items():
+        candidate = '{}:{}'.format(rack_value.ip_address, rack_value.partition)
+        rack_count_dict[candidate] = rack_count_dict.get(candidate, 0) + 1
+    return [k for k, v in rack_count_dict.items() if v > 1]
+
+
 def convert_to_rack_vthunder_conf(rack_list):
     """ Validates for all vthunder nouns for rack devices
         configurations.
     """
     rack_dict = {}
     validation_flag = False
-    try:
-        for rack_device in rack_list:
-            validation_flag = validate_params(rack_device)
-            if validation_flag:
-                rack_device['undercloud'] = True
-                vthunder_conf = data_models.VThunder(**rack_device)
-                rack_dict[rack_device['project_id']] = vthunder_conf
-            else:
-                LOG.warning('Invalid definition of rack device for '
-                            'project ' + rack_device['project_id'])
+    for rack_device in rack_list:
+        validation_flag = validate_params(rack_device)
+        if validation_flag:
+            rack_device['undercloud'] = True
+            if not rack_device.get('partition'):
+                rack_device['partition'] = a10constants.SHARED_PARTITION
+            vthunder_conf = data_models.VThunder(**rack_device)
+            if rack_dict.get(rack_device['project_id']):
```

Comment:
Let's make this the very first check. This way we don't spend cycles on turning it to a data model or adding partition if it's just going to fail anyway.